### PR TITLE
Qualcomm 2 16 merge

### PR DESCRIPTION
**Pull request by ronlieb**
Requesting review of Qualcomm changes from our local git repo to hexagon branch of halide.org.
Includes a merge from the hexagon branch of halide.org into our local git repo, prior to push.

### CODE REVIEW

File: src/Argument.h [hunk 1]
```diff
@@ -56,13 +56,11 @@ struct Argument {
      * By default, they are left unset, implying "no default, no min, no max". */
     Expr def, min, max;
 
-  int stride_multiples[4];
```
Comment (dsharletg) [inline src/Argument.h]: Although the end result looks correct, I don't understand why these changes are showing up as commits to this branch. This change should already been in the hexagon branch. Note that https://github.com/halide/Halide/compare/hexagon shows no changes to Argument.h relative to master. I'm really nervous there's something wrong with the branch. 

@zvookin @abadams @steven-johnson do an of you have a better sense of what is going on here? I'm not fully confident in my understanding of git.

**Comment by dsharletg** [comment]
Thanks for the PR!

I'm trying to figure out why there are still tons of commits that have been made to master showing up in this PR. There are also some mystery changes (see my comment on Argument.h for more info) that I'm trying to understand.

It looks like maybe the problem is that the hexagon branch on github was created by copying files out of an internal git instance on your end. This worked fine then. However, now you are trying to merge from that internal git instance that still has all the history, but isn't aware of the files that were copied into this branch at first. I think this might be a situation where you need to use git rebase, to rebase all your internal changes onto the github hexagon branch? I'm not an expert on this, I've only had to do it once.

I'm sorry to be a pain in the butt about this. I know git can be finicky, but I want to avoid a situation where our branch is permanently confused. I've gotten myself into similar trouble before.

File: src/Argument.h (hunk 1 above)
Comment (ronlieb) [inline src/Argument.h]: Is it possible that this is the first time we have pulled git commits from the Qualcomm Halide git repo via 'git merge' into the hexagon branch?  
Previously we did a 'big blob merge' which did not have all our history, and now we are getting the individual commits whose content matches what is there?

Comment (dsharletg) [inline src/Argument.h]: I think that is exactly what is happening. See my comment on the PR (not just the file). I think maybe this is a situation where you need to use git rebase to fix the history.

Again, sorry to be a pain about this.

Comment (ronlieb) [inline src/Argument.h]: i am fine with trying to get it right. But i am not sure what/how to rebase at this point, could you elaborate a bit?

Comment (dsharletg) [inline src/Argument.h]: I'm not an expert on this myself. The basic idea is to replay the changes you have made to a branch onto another branch. I think what you want to do is this:

git checkout your_internal_branch
git rebase hexagon  ; hexagon being the github.com/halide/Halide hexagon branch

That *might* just work.

However, you might also want to use git rebase -i hexagon, which will let you interactively replay each commit. If there are redundant changes, this might be useful to skip them. I'm not sure that will be necessary or not. I *think* maybe not.

I gathered this info from reading this page: https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase

Sorry I cannot be more precise with my suggestions. I'd need to try it with your branch to be more sure.

Comment (ronlieb) [inline src/Argument.h]: i will look into it some more this afternoon, thanks for the suggestions.

Comment (dsharletg) [inline src/Argument.h]: Thinking about this more: I think you do need to do git rebase -i, so you can skip all of the commits that are redundant (basically, all of the commits from anyone not at Qualcomm, and probably all the commits making changes that were already in the "big blob merge" as you call it.

Comment (ronlieb) [inline src/Argument.h]: Sounds like a good suggestion, thanks

Comment (dsharletg) [inline src/Argument.h]: So, I realized that I could attempt to do this rebase myself. I tried to do it, and it is very difficult. There are many, many conflicts.

I think the best thing to do here is to just make new commits manually for the changes you want, on a new branch starting from the github.com/halide/Halide hexagon branch. I don't think there actually are that many. I'll go through the diff on this PR and make comments on the changes that should *not* be made.

File: Makefile [hunk 2]
```diff
@@ -118,7 +118,7 @@ HEXAGON_CXX_FLAGS=$(if $(WITH_HEXAGON), -DWITH_HEXAGON=1, )
 HEXAGON_LLVM_CONFIG_LIB=$(if $(WITH_HEXAGON), hexagon, )
 
 CXX_WARNING_FLAGS = -Wall -Werror -Wno-unused-function -Wcast-qual -Wignored-qualifiers -Wno-comment -Wsign-compare
-CXX_FLAGS = $(CXX_WARNING_FLAGS) -fno-rtti -Woverloaded-virtual $(FPIC) $(OPTIMIZE) -g -fno-omit-frame-pointer -DCOMPILING_HALIDE $(BUILD_BIT_SIZE)
+CXX_FLAGS = $(CXX_WARNING_FLAGS) -fno-rtti -Woverloaded-virtual -fPIC $(OPTIMIZE) -g -fno-omit-frame-pointer -DCOMPILING_HALIDE $(BUILD_BIT_SIZE) -I ${SRC_DIR}/hvx_arch_dev -I${SRC_DIR}/hvx_arch_stubs
```
Comment (dsharletg) [inline Makefile]: The change to fPIC should be reverted (unchanged from master).

File: apps/camera_pipe/process.cpp [hunk 3]
```diff
@@ -105,6 +105,9 @@ int main(int argc, char **argv) {
         fprintf(stderr, "%4d ", _lut[i]);
     }
     fprintf(stderr, "\n");
+    Image_info(lut);
```
Comment (dsharletg) [inline apps/camera_pipe/process.cpp]: I think that this change (the image deubgging stuff) should be made as a separate PR to master (not the hexagon branch). It's independent of hexagon, and we'd like to suggest changes on it separately.

File: src/Argument.h [hunk 4]
```diff
@@ -56,13 +56,11 @@ struct Argument {
...
     Argument() : kind(InputScalar), dimensions(0) {}
     Argument(const std::string &_name, Kind _kind, const Type &_type, uint8_t _dimensions,
                 Expr _def = Expr(),
                 Expr _min = Expr(),
-                Expr _max = Expr(),
-                int *_stride_multiples = nullptr) :
```
Comment (dsharletg) [inline src/Argument.h]: These changes should not be happening, this file should be unchanged from hexagon.

File: src/CodeGen_LLVM.cpp [hunk 5]
```diff
@@ -552,23 +552,6 @@ void CodeGen_LLVM::compile_func(const LoweredFunc &f) {
         }
     }
 
-    for (size_t i = 0; i < args.size(); i++) {
```
Comment (dsharletg) [inline src/CodeGen_LLVM.cpp]: This should be unchanged from the hexagon branch.

File: src/CodeGen_LLVM.cpp [hunk 6]
```diff
@@ -1672,9 +1655,7 @@ void CodeGen_LLVM::add_tbaa_metadata(llvm::Instruction *inst, string buffer, Exp
 
     inst->setMetadata("tbaa", tbaa);
 }
-ModulusRemainder CodeGen_LLVM::getAlignmentInfo(Expr e) {
-  return modulus_remainder(e, alignment_info);
-}
+
```
Comment (dsharletg) [inline src/CodeGen_LLVM.cpp]: Should be unchanged from the hexagon branch.

File: src/CodeGen_LLVM.cpp [hunk 7]
```diff
@@ -1708,7 +1689,7 @@ void CodeGen_LLVM::visit(const Load *op) {
             // can infer based on the index alone.
 
             // Boost the alignment if possible, up to the native vector width.
-            ModulusRemainder mod_rem = modulus_remainder(ramp->base, alignment_info);
+            ModulusRemainder mod_rem = get_alignment_info(ramp->base);
```
Comment (dsharletg) [inline src/CodeGen_LLVM.cpp]: Should be unchanged from the hexagon branch.

File: src/CodeGen_LLVM.cpp [hunk 8]
```diff
@@ -3166,7 +3149,7 @@ void CodeGen_LLVM::visit(const Store *op) {
             int native_bytes = native_bits / 8;
 
             // Boost the alignment if possible, up to the native vector width.
-            ModulusRemainder mod_rem = modulus_remainder(ramp->base, alignment_info);
+            ModulusRemainder mod_rem = get_alignment_info(ramp->base);
```
Comment (dsharletg) [inline src/CodeGen_LLVM.cpp]: Should be unchanged from the hexagon branch.

File: src/CodeGen_LLVM.cpp [hunk 9]
```diff
@@ -3479,4 +3468,8 @@ std::pair<llvm::Function *, int> CodeGen_LLVM::find_vector_runtime_function(cons
     return std::make_pair<llvm::Function *, int>(NULL, 0);
 }
 
+ModulusRemainder CodeGen_LLVM::get_alignment_info(Expr e) {
```
Comment (dsharletg) [inline src/CodeGen_LLVM.cpp]: Should be unchanged from the hexagon branch.

File: src/CodeGen_LLVM.h [hunk 10]
```diff
@@ -424,7 +424,10 @@ class CodeGen_LLVM : public IRVisitor {
      * If there's no match, returns (NULL, 0).
      */
     std::pair<llvm::Function *, int> find_vector_runtime_function(const std::string &name, int lanes);
-    ModulusRemainder getAlignmentInfo(Expr e);
+
+    /** Get the result of modulus-remainder analysis for a given expr. */
```
Comment (dsharletg) [inline src/CodeGen_LLVM.h]: Should be unchanged from the hexagon branch.

File: src/IROperator.h [hunk 11]
```diff
@@ -738,7 +738,15 @@ inline Expr abs(Expr a) {
...
-        return a;
+        /* Issue a warning, but do not discard the abs. This is because
+           on Hexagon, we have instructions that deal with unsigned overflow
+           and underflow in two different ways. Removing the abs would mean
+           unsigned arithmetic that leads to overflow/underflow will wrap
+           around, which is the way C defines it. However, on Hexagon we
+           have the absdiff instruction that relies on getting the magnitude
+           of the overflow, if any. For instance, abs(a-b), where b > a and
+           a and b are unsigned values, is not the same as (a-b)*/
+        //        return a;
```
Comment (dsharletg) [inline src/IROperator.h]: Should be unchanged from the hexagon branch.

File: src/LLVM_Runtime_Linker.cpp [hunk 12]
```diff
@@ -248,9 +248,8 @@ llvm::DataLayout get_data_layout_for_target(Target target) {
         return llvm::DataLayout("e-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-p:32:32:32-v128:32:32");
     } else if (target.arch == Target::Hexagon) {
       return llvm::DataLayout(
-         "e-m:e-p:32:32:32-i64:64:64-i32:32:32-i16:16:16-i1:8:8"
-         "-f64:64:64-f32:32:32-v64:64:64-v32:32:32-a:0-n16:32");
-
+         "e-m:e-p:32:32:32-a:0-n16:32-i64:64:64-i32:32:32-i16:16:16-i1:8:8"
```
Comment (dsharletg) [inline src/LLVM_Runtime_Linker.cpp]: I suggest making this change now in a very small branch and making a PR.

File: src/Param.cpp [hunk 13]
```diff
@@ -124,7 +120,7 @@ Internal::Parameter OutputImageParam::parameter() const {
 }
 
 OutputImageParam::operator Argument() const {
-  return Argument(name(), kind, type(), dimensions(), Expr(), Expr(), Expr(), param.stride_multiples());
+  return Argument(name(), kind, type(), dimensions(), Expr(), Expr(), Expr());
```
Comment (dsharletg) [inline src/Param.cpp]: Should be unchanged from the hexagon branch.

File: src/Param.h [hunk 14]
```diff
@@ -255,7 +255,6 @@ class OutputImageParam {
      * vectorizing. Known strides for the vectorized dimension
      * generate better code. */
     EXPORT OutputImageParam &set_stride(int dim, Expr stride);
-    EXPORT OutputImageParam &set_stride_multiple(int dim, int stride);
```
Comment (dsharletg) [inline src/Param.h]: Should be unchanged from the hexagon branch.

File: src/Param.h [hunk 15]
```diff
@@ -188,7 +188,7 @@ class OutputImageParam {
 
     /** Is this an input or an output? OutputImageParam is the base class for both. */
     Argument::Kind kind;
-    
+
```
Comment (dsharletg) [inline src/Param.h]: Should be unchanged from the hexagon branch.

File: src/Parameter.cpp [hunk 16]
```diff
@@ -20,7 +20,6 @@ struct ParameterContents {
     Expr min_constraint[4];
     Expr extent_constraint[4];
     Expr stride_constraint[4];
-    int stride_multiple[4];
```
Comment (dsharletg) [inline src/Parameter.cpp]: Should be unchanged from the hexagon branch.

File: src/Parameter.cpp [hunk 17]
```diff
@@ -28,9 +27,6 @@ struct ParameterContents {
         // dense vectorization. You can unset it by setting it to a
         // null expression. (param.set_stride(0, Expr());)
         stride_constraint[0] = 1;
-        int i;
-        for (i = 0; i < 4; i++)
```
Comment (dsharletg) [inline src/Parameter.cpp]: Should be unchanged from the hexagon branch.

File: src/Parameter.cpp [hunk 18]
```diff
@@ -212,12 +208,6 @@ void Parameter::set_stride_constraint(int dim, Expr e) {
     contents.ptr->stride_constraint[dim] = e;
 }
 
-void Parameter::set_stride_multiple(int dim, int i) {
```
Comment (dsharletg) [inline src/Parameter.cpp]: Should be unchanged from the hexagon branch.

File: src/Parameter.cpp [hunk 19]
```diff
@@ -236,16 +226,6 @@ Expr Parameter::stride_constraint(int dim) const {
     return contents.ptr->stride_constraint[dim];
 }
 
-int Parameter::stride_multiple(int dim) const {
```
Comment (dsharletg) [inline src/Parameter.cpp]: Should be unchanged from the hexagon branch.

File: src/Parameter.h [hunk 20]
```diff
@@ -115,12 +115,9 @@ class Parameter {
     EXPORT void set_min_constraint(int dim, Expr e);
     EXPORT void set_extent_constraint(int dim, Expr e);
     EXPORT void set_stride_constraint(int dim, Expr e);
-    EXPORT void set_stride_multiple(int dim, int e);
```
Comment (dsharletg) [inline src/Parameter.h]: Should be unchanged from the hexagon branch.

File: src/Parameter.h [hunk 21]
```diff
@@ -115,12 +115,9 @@ class Parameter {
...
     EXPORT Expr min_constraint(int dim) const;
     EXPORT Expr extent_constraint(int dim) const;
     EXPORT Expr stride_constraint(int dim) const;
-    EXPORT int stride_multiple(int dim) const;
```
Comment (dsharletg) [inline src/Parameter.h]: Should be unchanged from the hexagon branch.

File: src/Pipeline.cpp [hunk 22]
```diff
@@ -329,8 +329,7 @@ class InferArguments : public IRGraphVisitor {
         InferredArgument a = {
             Argument(p.name(),
                      p.is_buffer() ? Argument::InputBuffer : Argument::InputScalar,
-                     p.type(), p.dimensions(), def, min, max, p.is_buffer() ?
-                     p.stride_multiples() : nullptr),
+                     p.type(), p.dimensions(), def, min, max),
```
Comment (dsharletg) [inline src/Pipeline.cpp]: Should be unchanged from the hexagon branch.

File: src/Pipeline.cpp [hunk 23]
```diff
@@ -521,11 +520,10 @@ Module Pipeline::compile_to_module(const vector<Argument> &args,
     // Add the output buffer arguments
     for (Function out : contents.ptr->outputs) {
         for (Parameter buf : out.output_buffers()) {
-          // FIXME: PDB: Patch this up for stride_multiples
             public_args.push_back(Argument(buf.name(),
                                            Argument::OutputBuffer,
                                            buf.type(), buf.dimensions(), Expr(),
-                                           Expr(), Expr(), buf.stride_multiples()));
+                                           Expr(), Expr()));
```
Comment (dsharletg) [inline src/Pipeline.cpp]: Should be unchanged from the hexagon branch.

File: src/Simplify.cpp [hunk 24]
```diff
@@ -2821,7 +2821,9 @@ class Simplify : public IRMutator {
                 expr = make_const(op->type, ia);
             } else if (ta.is_uint()) {
               // abs(uint) is a no-op.
-              expr = a;
+              // For Hexagon, do not treat this as a no-op. See explanation
```
Comment (dsharletg) [inline src/Simplify.cpp]: Should be unchanged from the hexagon branch.

File: test/hexagon/include/halide-hexagon-setup.h [hunk 25]
```diff
@@ -58,12 +61,11 @@ void set_min(ImageParam &I, int dim, Expr a) {
 void set_output_buffer_min(Func &f, int dim, Expr a) {
   f.output_buffer().set_min(dim, a);
 }
-void set_stride_multiple(ImageParam &I, int dim, int m) {
-  I.set_stride_multiple(dim, m);
+void set_stride_multiple(OutputImageParam I, int dim, int m) {
+  I.set_stride(dim, (I.stride(dim) / m) * m);
```
Comment (dsharletg) [inline test/hexagon/include/halide-hexagon-setup.h]: Should be unchanged from the hexagon branch.

File: test/hexagon/include/halide-hexagon-setup.h [hunk 26]
```diff
@@ -58,12 +61,11 @@ void set_min(ImageParam &I, int dim, Expr a) {
...
 }
-void set_stride_multiple(Func &f, int dim, int m) {
-  Expr stride = f.output_buffer().stride(dim);
-  f.output_buffer().set_stride_multiple(dim, m);
+void set_stride_multiple(Func f, int dim, int m) {
```
Comment (dsharletg) [inline test/hexagon/include/halide-hexagon-setup.h]: Should be unchanged from the hexagon branch.

File: tools/halide_image_info.h [hunk 27]
```diff
@@ -0,0 +1,333 @@
+// This header defines a several methods useful for debugging programs that
```
Comment (dsharletg) [inline tools/halide_image_info.h]: I'd like it if this change was made in a separate branch from master, not the hexagon branch.

**Comment by dsharletg** [comment]
To summarize the comments I just made: I attempted to rebase this onto the hexagon branch, it was extremely difficult. 

I think the easiest thing to do is just make the changes manually in a new branch starting from the github.com/halide/Halide hexagon branch. Therefore, I marked all of the changes that *should not* happen. 

The remaining changes are actually relatively small: just a few changes to CodeGen_Hexagon, and lots of changes to the makefiles/tests. It's because these changes are not huge/intertwined with the rest of the code, that I think it will be easiest to just make these commits manually.

**Comment by ronlieb** [comment]
Thanks for the detailed suggestions and comments, and yes it looked very painful indeed.
I think your most recent suggestion to start clean with hexagon branch and push change sets up is a good one , so ...
i am going to proceed with manual pushes of related changes (creating new pull requests)
1) data_layout
2) test makefile changes
3) relevant CodeGen_Hexagon* changes

i will

**Comment by dsharletg** [comment]
Thanks so much for doing this. I just read this message after addressing the other PR, and I see you intended to make 3 separate PRs. I think that's great, we'd like to address things in smaller pieces. But, it came through as one large PR. The problem is likely that when you push commits to a branch that is open for a PR, those commits become part of that PR. To avoid this, you'll need to make 3 separate branches. But, I wouldn't worry about it for this set of changes, this is just FYI for future changes. 

One possibly useful tip is if you have several dependent changes that you want to make as separate PRs, you can branch off of one of your change branches to another, and continue editing. The workflow for this would look like:

git checkout hexagon   ; start at hexagon branch
git checkout -b feature_1 ; make a new branch for feature 1
git commit ...   ; add some commits to feature 1
git push https://github.com/halide/Halide feature_1  ; push feature_1 to github, can be made into a PR
git checkout -b feature_2  ; make a new branch for feature 2, starting where feature_1 left off
git commit ...  ; add some commits to feature 2
git push https://github.com/halide/Halide feature_2  ; push feature_2 to github, can be made into a PR

If changes to feature_1 happen, you can merge them into the feature_2 branch with little trouble:
git checkout feature_2
git merge feature_1

However, note that this probably isn't necessary for this particular set of changes, because they look to be independent.

I'm going to close this PR, since it is superceded by #1035. Thanks again for fixing the merge issues!

**Comment by ronlieb** [comment]
thanks Dillon. i can look at breaking them up into separate PRs.

**Comment by dsharletg** [comment]
Don't worry about it for this round of changes! It's not a big deal, I've already read through the PR.
